### ISSUE TITLE
`msiGetObjType` returns `-1105000` when called by rodsusers against the root collection "/"

### ISSUE THREAD

**Issue by mstfdkmn** [irods/irods#7983]
### iRODS Version, OS and Version
4.3.2, almalinux9
iCommands 4.3.2

4.3.3, almalinux9
iCommands 4.3.3

### What did you try to do?
In a custom PEP rule, I attempted to use `msiGetObjType` to determine the object type. When a user removed its own acl, the mentioned msi is failing.

```
[mustafa@icts-t-dev-u0137480-irods ~]$ ils -A /test/home/mustafa/test.txt
  /test/home/mustafa/test.txt
        ACL - mustafa#test:own

[mustafa@icts-t-dev-u0137480-irods ~]$ ichmod null mustafa test.txt
remote addresses: 10.89.0.10 ERROR: rcModAccessControl failure  status = -1105000 INVALID_OBJECT_TYPE
```

```
def pep_api_mod_access_control_post(rule_args, callback, rei):
    path = str(rule_args[2].path)
    obj_type = callback.msiGetObjType(path, "")["arguments"][1]
```

### Expected behavior
`msiGetObjType` should function correctly for any given path, regardless of the granted permissions. Or at least for the root paths such as "/", "/zoneName" and "/zoneName/home"

### Observed behavior (including steps to reproduce, if applicable)
When it is called by rodsusers it fails with:

logs:
`{"processname": "irods", "groupname": "irods", "pid": "151", "channel": "stdout", "irods": {"log_category": "rule_engine", "log_level": "error", "log_message": "caught python exception", "python_exception": "Traceback (most recent call last):\n  File \"/etc/irods/helpers.py\", line 20, in inner\n    result = f(*args, **kwargs)\n  File \"/etc/irods/rules_global.py\", line 45, in pep_api_mod_access_control_post\n    obj_type = callback.msiGetObjType(path, '')[\"arguments\"][1]\nRuntimeError: [iRods__Error__Code:-1105000] [-]\t/irods_source/server/re/include/irods/irods_re_plugin.hpp:333:irods::error irods::dynamic_operation_execution_manager<std::tuple<>, RuleExecInfo *, irods::DONT_AUDIT_RULE>::call(const std::string &, const std::string &, OP, As &&...) [T = std::tuple<>, C = RuleExecInfo *, Audit = irods::DONT_AUDIT_RULE, OP = std::function<irods::error (const std::string &, RuleExecInfo *&, irods::unpack &&)>, As = <const std::string &, RuleExecInfo *&, irods::unpack>] :  status [INVALID_OBJECT_TYPE]  errno [] -- message [exec_microservice_adapter failed]\n\t[-]\t/irods_source/server/re/src/irods_re_plugin.cpp:133:irods::error irods::default_microservice_manager<RuleExecInfo *>::exec_microservice_adapter(std::string, irods::default_ms_ctx, std::list<boost::any> &) :  status [INVALID_OBJECT_TYPE]  errno [] -- message [exec_microservice_adapter failed]\n\n\n", "request_api_name": "MOD_ACCESS_CONTROL_AN", "request_api_number": 707, "request_api_version": "d", "request_client_user": "mustafa", "request_host": "10.89.0.10", "request_proxy_user": "mustafa", "request_release_version": "rods4.3.2", "rule_engine_plugin": "python", "server_host": "icts-t-dev-u0137480-irods.service.qsvcd", "server_pid": 7351, "server_timestamp": "2024-10-23T13:11:46.151Z", "server_type": "agent", "server_zone": "test"}}`

to reproduce easily as a rodsuser, you can run the following snippet  as well:
```
def main(rule_args, callback, rei):
    callback.writeLine("serverLog", "Test...")
    path = "/yourZone/home/rodsUser/test.txt"
    obj_type = callback.msiGetObjType(path, "")["arguments"][1]
    callback.writeLine("serverLog", f"{obj_type}")
    root_path = "/"
    obj_type = callback.msiGetObjType(root_path, "")["arguments"][1]
    callback.writeLine("serverLog", f"{obj_type}")

INPUT null
OUTPUT ruleExecOut

```

**Comment by alanking** [comment]
The `ichmod` being performed here fully explains what is happening, and this is the expected behavior.

TLDR: `rodsuser`s need `read` permission (or higher) to stat things. Try changing the permission to `read` or `write` instead of `null` and the example PEP given works fine for `rodsuser`s.

The object only has one permission set: `own` permission for user `mustafa`. Changing the permission to `null` for user `mustafa` means that no users have permissions on that object after the `ichmod`. The stat performed in the post-PEP fails because the user executing the PEP has no permission on the object anymore - this is the expected behavior. `rodsadmin`s can stat anything, regardless of permissions, which is why this problem does not manifest for `rodsadmin`s.

"/", "/zoneName", etc. normally have no permissions set for regular users (at least in a default setup), so naturally the same problem occurs with these as in the example data object with no permissions set. Here's what mine looks like (again, pretty vanilla zone):
```
$ ils -A / /tempZone
/:
        ACL - rods#tempZone:own   
        Inheritance - Disabled
  C- /tempZone
/tempZone:
        ACL - rods#tempZone:own   
        Inheritance - Disabled
  C- /tempZone/home
  C- /tempZone/trash
```
Only `rods` (a `rodsadmin`) has any permission on "/" and "/tempZone" and "/tempZone/home" in my setup. Therefore, any non-`rodsadmin` user will not be able to stat those collections. You must give users permission to be able to stat those collections.

If you want all users to be able to stat root collection (or anything else for that matter), you will need to grant at least `read` permission on it (maybe there's something more restrictive that still allows this, but I'm not sure).

Finally, the rule snippet at the end cannot be executed by `rodsuser`s because `rodsuser`s are not allowed to run python rules directly via `irule`. See https://github.com/irods/irods_rule_engine_plugin_python/issues/105.

Hope that helps.

**Comment by mstfdkmn** [comment]
Hi @alanking,

Yes, your update clarifies everything. However, in our audit flow, we need to capture any actions related to data objects. To achieve this, we have developed custom rules that correctly capture ACL changes and send those messages to a broker.

The reason we have a custom rule for ACL change operations, in addition to the audit plugin, is due to this issue. Specifically, we need to capture when the last remaining ACL is removed (i.e., when null is given) from a data object. I believe I can handle this with this pull request https://github.com/irods/irods_rule_engine_plugin_python/pull/228. I will keep you updated.


> Finally, the rule snippet at the end cannot be executed by rodsusers because rodsusers are not allowed to run python rules directly via irule. See https://github.com/irods/irods_rule_engine_plugin_python/issues/105.

That is my bad! I posted the one for the PREP by mistake.. The one below should have been posted:

```
myRule{
    writeLine("serverLog", "Test...")
    *path = "/test/home/mustafa/test.txt"
    msiGetObjType(*path, *Type)
    writeLine("serverLog", *Type)
    *root_path = "/"
    msiGetObjType(*root_path, *Type)
    writeLine("serverLog", *Type)
}

INPUT null
OUTPUT ruleExecOut
```

**Comment by alanking** [comment]
Ah, I see. Thanks.

Have you considered stashing away the object type in a pre-PEP (before the permission is set to `null`) and then fetch it back out in the post-PEP (after the permission is successfully set to `null`) to log it rather than actually performing the stat in the post-PEP? This would follow the recipe described here: https://docs.irods.org/4.3.3/administrators/policy_cookbook/#sharing-data-across-peps

**Comment by mstfdkmn** [comment]
That seems like a straightforward solution, which I hadn’t considered. I’ve implemented it and am currently testing it in our test tier. However, it appears that sharing object type data across PEPs (from pre to post) is a bit slowing down the client operation, especially when the operation is recursive for a large collection. Do you think this is to be expected? For context, we construct a message (after querying data objects in the given collection) and send it to a broker. 

Btw, please note that we conducted extensive stress tests when all logic was placed in the post PEP. And all had gone well.

**Comment by alanking** [comment]
When you say you are constructing a message and sending it to a broker, is that to share the data object type between the PEPs, or is that for logging / auditing purposes?

If the latter, it is a bit surprising that splitting the work across two PEPs is significantly slower than doing everything just in the post-PEP. If you're able to share more details about this, that could be helpful for understanding the slowdown. The only difference - in theory - should be that the query is being done in the pre-PEP and the results being logged in the post-PEP. The overhead of calling the pre-PEP should not be very high.

If the former, given that the PEPs should be executing inside a single agent, you should be able to use a persistent in-memory variable. No need for sending/receiving the information from a broker.

**Comment by mstfdkmn** [comment]
Thanks, Alan. It’s the latter. I’ll conduct more systematic tests next week. In the post above, I shared my observations from the manual test I did.

```
@pep
def pep_api_mod_access_control_pre(rule_args, callback, rei):
    global obj_type
    obj_type = callback.msiGetObjType(str(rule_args[2].path), "")["arguments"][1]


@pep
def pep_api_mod_access_control_post(rule_args, callback, rei):

    path = str(rule_args[2].path)
    # obj_type = callback.msiGetObjType(path, "")["arguments"][1]
    # obj_type should come from pre pep to bypass -1105000 in case a last remaining acl is removed
    if obj_type == "-c" and rule_args[2].recursiveFlag == 0:
        return
    if obj_type == "-d":
        acl_message = make_message(path, rule_args, rei)
        p = create_kafka_producer(callback)
        produce_acl_message(acl_message, p)
        p.flush()
        return
    if obj_type == "-c" and rule_args[2].recursiveFlag == 1:
        p = create_kafka_producer(callback)
        for items in get_data_objects(path, callback):
            data_obj_path = f"{items[0]}/{items[1]}"
            acl_message = make_message(data_obj_path, rule_args, rei)
            produce_acl_message(acl_message, p)
        p.flush()
        return

```

**Comment by alanking** [comment]
Just to highlight how I imagine the diff...
```diff
+ @pep
+ def pep_api_mod_access_control_pre(rule_args, callback, rei):
+     global obj_type
+     obj_type = callback.msiGetObjType(str(rule_args[2].path), "")["arguments"][1]


 @pep
 def pep_api_mod_access_control_post(rule_args, callback, rei):

     path = str(rule_args[2].path)
-    obj_type = callback.msiGetObjType(path, "")["arguments"][1]
+    # obj_type = callback.msiGetObjType(path, "")["arguments"][1]
+    # obj_type should come from pre pep to bypass -1105000 in case a last remaining acl is removed
     if obj_type == "-c" and rule_args[2].recursiveFlag == 0:
         return
     if obj_type == "-d":
         acl_message = make_message(path, rule_args, rei)
         p = create_kafka_producer(callback)
         produce_acl_message(acl_message, p)
         p.flush()
         return
     if obj_type == "-c" and rule_args[2].recursiveFlag == 1:
         p = create_kafka_producer(callback)
         for items in get_data_objects(path, callback):
             data_obj_path = f"{items[0]}/{items[1]}"
             acl_message = make_message(data_obj_path, rule_args, rei)
             produce_acl_message(acl_message, p)
         p.flush()
         return
```
The only difference is that the query is happening at an earlier time and the `obj_type` is a global. In theory, this should have no impact on runtime because the same amount of work is being done.

Yes, please let us know what you find in your testing. Thanks

**Comment by mstfdkmn** [comment]
Hi Alan, I believe we can close this issue as my tests haven’t yielded any meaningful results so far. Thanks for your understanding.

**Comment by alanking** [comment]
Excellent. Thanks
